pkg/traceql/engine_metrics_average.go

```diff
@@ -0,0 +1,540 @@
+		return
+	}
+	inc := g.getSpanAttValue(span)
+	if math.IsNaN(inc) {
+		return
+	}
+	s.count[interval]++
+	mean, c := averageInc(s.avg[interval], inc, s.count[interval], s.compensation[interval])
+	s.avg[interval] = mean
+	s.compensation[interval] = c
+}
+
+func averageInc(mean, inc, count, compensation float64) (float64, float64) {
+	if math.IsNaN(mean) && !math.IsNaN(inc) {
+		// When we have a proper value in the span we need to initialize to 0
+		mean = 0
+	}
+	if math.IsInf(mean, 0) {
+		if math.IsInf(inc, 0) && (mean > 0) == (inc > 0) {
+			// The `current.val` and `new` values are `Inf` of the same sign.  They
+			// can't be subtracted, but the value of `current.val` is correct
+			// already.
+			return mean, compensation
+		}
+		if !math.IsInf(inc, 0) && !math.IsNaN(inc) {
+			// At this stage, the current.val is an infinite. If the added
+			// value is neither an Inf or a Nan, we can keep that mean
+			// value.
+			// This is required because our calculation below removes
+			// the mean value, which would look like Inf += x - Inf and
+			// end up as a NaN.
+			return mean, compensation
+		}
+	}
+	mean, c := kahanSumInc(inc/count-mean/count, mean, compensation)
+	return mean, c
+}
+
+func kahanSumInc(inc, sum, c float64) (newSum, newC float64) {
+	t := sum + inc
+	switch {
+	case math.IsInf(t, 0):
+		c = 0
+
+	// Using Neumaier improvement, swap if next term larger than sum.
+	case math.Abs(sum) >= math.Abs(inc):
+		c += (sum - t) + inc
+	default:
+		c += (inc - t) + sum
+	}
+	return t, c
+}
+
+func (g *AvgOverTimeSpanAggregator[F, S]) ObserveExemplar(span Span, value float64, ts uint64) {
+	if !g.getGroupingValues(span) {
+		return
+	}
+
+	// Observe exemplar
+	all := span.AllAttributes()
+	lbls := make(Labels, 0, len(all))
+	for k, v := range span.AllAttributes() {
+		lbls = append(lbls, Label{k.String(), v})
+	}
+	s := g.getSeries()
+
+	if s.exemplarBuckets.testTotal() {
+		return
+	}
+	interval := IntervalOfMs(int64(ts), g.start, g.end, g.step)
+	if s.exemplarBuckets.addAndTest(interval) {
+		return
+	}
+
+	s.exemplars = append(s.exemplars, Exemplar{
+		Labels:      lbls,
+		Value:       value,
+		TimestampMs: ts,
+	})
+	g.series[g.buf.fast] = s
+}
+
+func (g *AvgOverTimeSpanAggregator[F, S]) labelsFor(vals S, t string) (Labels, string) {
+	if g.by == nil {
+		serieLabel := make(Labels, 1, 2)
+		serieLabel[0] = Label{labels.MetricName, NewStaticString(metricsAggregateAvgOverTime.String())}
+		if t != "" {
+			serieLabel = append(serieLabel, Label{"_type", NewStaticString(t)})
+		}
+		return serieLabel, serieLabel.String()
+	}
+	labels := make(Labels, 0, len(g.by)+1)
+	for i := range g.by {
+		if vals[i].Type == TypeNil {
+			continue
+		}
+		labels = append(labels, Label{g.by[i].String(), vals[i]})
+	}
+
+	if len(labels) == 0 {
+		// When all nil then force one
+		labels = append(labels, Label{g.by[0].String(), NewStaticNil()})
+	}
+
+	if t != "" {
+		labels = append(labels, Label{"_type", NewStaticString(t)})
```

Comment:
Let's move these strings to consts.   `Compare()` is using `internalLabelMetaType`, can we use that here too?  It's currently in engine_metrics_compare.go but we can relocate it.  For `_count` what about `internalMetaTypeCount`?